jxl/src/container/mod.rs

```diff
@@ -0,0 +1,261 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ContainerParser")
+            .field("state", &self.state)
+            .field("next_jxlp_index", &self.next_jxlp_index)
+            .finish_non_exhaustive()
+    }
+}
+
+#[derive(Debug, Default)]
+enum DetectState {
+    #[default]
+    WaitingSignature,
+    WaitingBoxHeader,
+    WaitingJxlpIndex(ContainerBoxHeader),
+    InAuxBox {
+        header: ContainerBoxHeader,
+        data: Vec<u8>,
+        bytes_left: Option<usize>,
+    },
+    InCodestream {
+        kind: BitstreamKind,
+        bytes_left: Option<usize>,
+    },
+    Done(BitstreamKind),
+}
+
+/// Structure of the decoded bitstream.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum BitstreamKind {
+    /// Decoder can't determine structure of the bitstream.
+    Unknown,
+    /// Bitstream is a direct JPEG XL codestream without box structure.
+    BareCodestream,
+    /// Bitstream is a JPEG XL container with box structure.
+    Container,
+    /// Bitstream is not a valid JPEG XL image.
+    Invalid,
+}
+
+impl ContainerParser {
+    const CODESTREAM_SIG: [u8; 2] = [0xff, 0x0a];
+    const CONTAINER_SIG: [u8; 12] = [0, 0, 0, 0xc, b'J', b'X', b'L', b' ', 0xd, 0xa, 0x87, 0xa];
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn kind(&self) -> BitstreamKind {
+        match self.state {
+            DetectState::WaitingSignature => BitstreamKind::Unknown,
+            DetectState::WaitingBoxHeader
+            | DetectState::WaitingJxlpIndex(..)
+            | DetectState::InAuxBox { .. } => BitstreamKind::Container,
+            DetectState::InCodestream { kind, .. } | DetectState::Done(kind) => kind,
+        }
+    }
+
+    pub fn feed_bytes(&mut self, input: &[u8]) -> Result<(), Error> {
+        let state = &mut self.state;
+        let buf = &mut self.buf;
+        buf.extend_from_slice(input);
+
+        loop {
+            match state {
+                DetectState::WaitingSignature => {
+                    if buf.starts_with(&Self::CODESTREAM_SIG) {
+                        // tracing::debug!("Codestream signature found");
+                        *state = DetectState::InCodestream {
+                            kind: BitstreamKind::BareCodestream,
+                            bytes_left: None,
+                        };
+                        continue;
+                    }
+                    if buf.starts_with(&Self::CONTAINER_SIG) {
+                        // tracing::debug!("Container signature found");
+                        *state = DetectState::WaitingBoxHeader;
+                        buf.drain(..Self::CONTAINER_SIG.len());
+                        continue;
+                    }
+                    if !Self::CODESTREAM_SIG.starts_with(buf)
+                        && !Self::CONTAINER_SIG.starts_with(buf)
+                    {
+                        // tracing::error!("Invalid signature");
```

Comment:
Feel free to uncomment those, but I would only use DEBUG/TRACE levels so they are omitted by default in builds.
(I believe there is no way to set the tracing static log level per-crate)

Reply:
https://users.rust-lang.org/t/turn-off-tracing-at-compile-time-for-specific-crate/104337/2

actually, maybe this is a better way... (feel free to do it in a separate PR)

Reply:
I'd prefer doing that in a separate PR.

Could you merge this if it looks good to you? The branch protection rule doesn't seem to allow me to merge PRs...